### DISCUSSION TITLE
Question: `DeclareWide` vs `StringW`

### DISCUSSION THREAD

**Discussion by sancarn** [twinbasic/twinbasic#1735] [General]
More of a question than a feature request. Just wondered what the logic was behind using the `DeclareWide` syntax vs having a seperate type for unicode/ascii strings (with `String` still being supported but non-explicit) in DLLs. I.E.

Instead of this:

```vb
Public Declare PtrSafe Function AbortSystemShutdownA Lib "advapi32"  (ByVal lpMachineName As String) As Long
Public DeclareWide PtrSafe Function AbortSystemShutdownW Lib "advapi32"  (ByVal lpMachineName As String) As Long
```

Why wasn't something like this chosen:

```vb
Public Declare PtrSafe Function AbortSystemShutdownA Lib "advapi32"  (ByVal lpMachineName As String) As Long
Public Declare PtrSafe Function AbortSystemShutdownW Lib "advapi32" (ByVal lpMachineName As StringW) As Long
```

?

----

I ask mainly because I believe introduction of `PtrSafe` by microsoft was a mistake and lead to loads of unnecessary conditional compilation. And although I can't see why this change would lead to a bunch of required conditional compilcation necessarily it kinda feels like the same sort of thing and a better option would be type aliases... I.E.

```vb
#if Unicode then
  type StringX as StringW
#else
  type StringX as StringA
#end if
Public Declare PtrSafe Function AbortSystemShutdownA Lib "advapi32"  (ByVal lpMachineName As StringX) As Long
```

Maybe I'm wrong though, and there is more of a need for seperate definitions.

**Comment by sancarn** [comment]
Hmm on 2nd inspection I do see an issue with using type aliases... you would need declare name aliases too...

```vb
#if Unicode then
  type StringX as StringW
  const sAbortSystemShutdown as string = "AbortSystemShutdownW"
#else
  type StringX as StringA
  const sAbortSystemShutdown as string = "AbortSystemShutdownA"
#end if
Public Declare PtrSafe Function AbortSystemShutdown Lib "advapi32" alias sAbortSystemShutdown (ByVal lpMachineName As StringX) As Long
```

**Comment by WaynePhillipsEA** [comment]
Symbols like `StringW` can be user-defined as types or classes in user code in existing codebases.  When adding new functionality there is a general preference for choosing something that is completely invalid in existing VB6/VBA codebases, so that there is no chance of conflicts.

**Reply by sancarn** [reply]
Hmm OK, I can see the benefits there, and thanks for explaining your reasoning 😁 I feel that generality is best, but it's all tradeoffs in the end.

**Reply by WaynePhillipsEA** [reply]
No problem.   The goal of 100% back-compatibility with VB6 (and VBA) trumps pretty much everything else.  Careful design choices around new features is key to maintaining that.  When introducing a new feature, we'll always look for a way to do it without breaking back-compat.

**Comment by fafalone** [comment]
Granted I don't spend a *ton* of time in the VBA world, but my experience has been there's tons of superfluous conditional compilation.

`Public Declare PtrSafe Function AbortSystemShutdownW Lib "advapi32"  (ByVal lpMachineName As LongPtr) As Long`

This is valid in both 32bit and 64bit. You only need a version without `PtrSafe` to support VB6 or pre-2010 Office. But I see tons of code entirely dependent on default object model presence (i.e. not designed to run in VB6) that not only does separate declares, but does separate variables too-- even though you can trivially add `LongPtr` as an alias to VB6 and pre-2010 Office (which is what I do with all new VB6 code now). 

---

More on the point here... the thing is, VB/tB strings are *already* Unicode. What's happening with APIs is behind the scenes, VB creates a temporary ANSI string out of the original Unicode string, passes that to the API, then converts it back to Unicode. So a new StringW type, even as an alias, would be very confusing... it would mean 'An identical string, which when used in APIs disables ANSI<->Unicode conversion'.... I think `DeclareWide` was a much better solution. 

In WinDevLib (formerly tbShellLib), most APIs are provided as a set:

```vba
Public Declare PtrSafe Function AbortSystemShutdownA Lib "advapi32" (ByVal lpMachineName As String) As BOOL
Public Declare PtrSafe Function AbortSystemShutdownW Lib "advapi32" (ByVal lpMachineName As LongPtr) As BOOL
Public DeclareWide PtrSafe Function AbortSystemShutdown Lib "advapi32" Alias "AbortSystemShutdownW" (ByVal lpMachineName As String) As BOOL
```

Though for some areas, I've been forgoing the ANSI version, since it's really bad practice to be writing new code with them, and the `DeclareWide` option has the same benefit of not needing `StrPtr`.

**Reply by sancarn** [reply]
Hmm I disagree that `DeclareWide` is "less confusing", ultimately it is just "new and unknown".  I understand the mechanism but realistically VB6/VBAs string type is the root of the confusion. IMO being able to declare exactly what you want both in memory and in dll calls (with ability to auto cast between them) is a better solution which also helps declares intent better... But that's just my opinion, each entitled to their own 😊

P.s. as far as I know there is no VBA7 32 bit?

**Reply by fafalone** [reply]
But it's the same in memory... that's why it would be confusing. It's not a different type; it's the exact same bytes of data. IMO since all you're doing is altering the behavior of the API call, it makes sense to modify the API call, not create a data type whose only purpose is to be 100% identical in memory layout but whose type name signals the API call to behave differently. Also saves a ton of string copying operations. But yeah, lots of different ways to see things.

---

There's still 32bit versions of Office. From 2010 on, they use VBA7. The key difference is that while you can use `PtrSafe` and `LongPtr`, you *can't* use `LongLong` in 32bit VBA7. Of course, if it was VBA6, you couldn't use any of those.

**Reply by sancarn** [reply]
> But it's the same in memory... that's why it would be confusing.

Huh? Implementing a `StringA` type would not be the same in memory? Only reason for `StringW` would be the FFI calls...? Or maybe I'm misunderstanding?

> IMO since all you're doing is altering the behavior of the API call, it makes sense to modify the API call

I agree if that was all we were doing, but no, my intention in OP was for `StringA` to literally implement an Ascii string in memory, as well as strictly defining use of them in FFI calls. Ofc, `String` itself would still be the same funky type for backwards compatibility.

----

> There's still 32bit versions of Office. From 2010 on, they use VBA7. The key difference is that while you can use PtrSafe and LongPtr, you can't use LongLong in 32bit VBA7.

Oh ofc you are right. This is why we use `Win64` comp constant when defining ptrsize etc. Silly me.

**Reply by fafalone** [reply]
I had assumed StringW was a Unicode string type, since you were using regular String (already Unicode) with AbortSystemShutdownA:

```vba
Public Declare PtrSafe Function AbortSystemShutdownA Lib "advapi32"  (ByVal lpMachineName As String) As Long
Public Declare PtrSafe Function AbortSystemShutdownW Lib "advapi32" (ByVal lpMachineName As StringW) As Long
```

Yes, I do agree we need an ANSI string type. But then you'd instead have

`
```vba
Public Declare PtrSafe Function AbortSystemShutdownA Lib "advapi32"  (ByVal lpMachineName As StringA) As Long
Public DeclareWide PtrSafe Function AbortSystemShutdownW Lib "advapi32" (ByVal lpMachineName As String) As Long
```

Which wouldn't be any different besides avoiding the internal conversion between them, if the language also implemented special handling for it. It wouldn't change the need to specify a way to turn off the conversion for regular Unicode `String` without breaking backcompat or creating a second Unicode string type (presumably we'd now have 3... StringA, String, and StringW (same as String except triggers different API behavior).

**Reply by sancarn** [reply]
Ah true! You are right I wasn't very clear what the suggestion I was making actually was 😂😅

> presumably we'd now have 3... `StringA`, `String`, and `StringW`

My thoughts exactly. At which point no particular reason to have `DeclareWide` anymore either(?), assuming `StringA` and `StringW` were unaffected DLL calls (though I do take Wayne's point about preventing compilation). It would also satisfy the unusual case where you'd have a Unicode string as 1 param and an ascii string as another param (though I don't know if this is realistic at all):

```vb
Public Declare PtrSafe Function ConcatAsciiToUnicode Lib "stringStuff" (ByVal a as StringW, ByVal b as StringA) as StringW
```

Which I believe isn't doable with `DeclareWide` syntax?

**Reply by wqweto** [reply]
IMO the original dillema was whether to introduce new `DeclareWide` vs new arg attribute like `ByWide`.

Currently in VBx they use `ByVal` on strings in an API declare to indicate ANSI<->Unicode conversion is required. It makes sense to introduce similar new attribute (for strings only) which would allow "converting" BSTRs to LPWSTR. It is very rare for an API to expect BSTRs so although similar the LPWSTR type is different although StrPtr "conversion" works in 99.9% of the cases.

I think `DeclareWide` won over `ByWide` (or similar) because it's shorter to impl for declares with multiple string parameters.

I was wondering if LPWSTR/LPSTR can be implemented as some kind of aliases (when aliases get introduced) so that these when being used in API declares signal ANSI vs Unicode transcoding is required (or nor).

**Reply by fafalone** [reply]
There's ByRef strings too though... it's even more rare than [in] BSTR, but [out] BSTR* definitely exists. 

Probably could have done it with an attribute. 

`Declare Sub foo Lib "bar" ([Unicode] ByVal lpwstr As String)`


@sancarn, you example is possible without DeclareWide... 

`Public Declare PtrSafe Function ConcatAsciiToUnicode Lib "stringStuff" (ByVal a as LongPtr, ByVal b as String) as LongPtr`

But I still think it would be a very confusing situation, especially for novices, that added tons of extra performance-zapping string operations to have two different Unicode string types. Unless we're going to go all-in and allow true LPSTR/LPWSTR so we can receive LPSTR*/LPWSTR* without dereferencing a pointer ourselves.
